apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/functions/AllocateAndAddFloatingIpToNode.java

```diff
@@ -156,8 +215,57 @@ public boolean apply(FloatingIP arg0) {
       return Optional.fromNullable(ip);
    }
 
+   private Optional<org.jclouds.openstack.neutron.v2.domain.FloatingIP> tryFindExistingFloatingIp(org.jclouds.openstack.neutron.v2.features.FloatingIPApi neutronFloatingApi, final String availabilityZone) {
+      Optional<org.jclouds.openstack.neutron.v2.domain.FloatingIP> floatingIPOptional = neutronFloatingApi.list().concat().firstMatch(new Predicate<org.jclouds.openstack.neutron.v2.domain.FloatingIP>() {
+         @Override
+         public boolean apply(@Nullable org.jclouds.openstack.neutron.v2.domain.FloatingIP input) {
+            return input.getPortId() == null && input.getAvailabilityZone().equals(availabilityZone);
+         }
+      });
+      return floatingIPOptional;
+   }
+
+   private org.jclouds.openstack.neutron.v2.domain.FloatingIP createFloatingIpUsingNeutron(org.jclouds.openstack.neutron.v2.features.FloatingIPApi neutronFloatingApi, NodeMetadata node, final String availabilityZone) {
+      String regionId = node.getLocation().getParent().getId();
+
+      Optional<Network> networkOptional = getNetworkApi(regionId).list().concat().firstMatch(new Predicate<Network>() {
+         @Override
+         public boolean apply(@Nullable Network input) {
+            return input.getExternal() && input.getAvailabilityZone().equals(availabilityZone);
+         }
+      });
+      if (!networkOptional.isPresent())
+         throw new InsufficientResourcesException("Failed to find a suitable external network.");
```

Comment:
To be more close to the old behavior, instead of looking for *any* external network, could we use the existing `poolNames` configuration to filter external networks by name? This way users will configure the same name `where` they want the floating IPs from, regardless of the API they are using.

Reply:
good point! as poolnames are optional, what shall we do if poolnames is empty? fail seems reasonable as we try to re-use an existing floating ip and to create from a poolname: if no poolnames are specified we can't infer much, right?

Reply:
If pool names is empty I'd say it is ok to pick the first external network we find.  We should add proper logs about the network we are selecting.